### ISSUE TITLE
Applying Mixout in Pretraind LM Models

### ISSUE THREAD

**Issue by Shashi456** [bloodwass/mixout#1]
Hello,
Could you explain how this code could be applied to pretrained language models?

**Comment by bloodwass** [comment]
Hi,
Mixout needs to replace some parameters of each _torch.nn.Module_, so you must substitute a mixout module for each original module of your pretrained model. I only offered _MixLinear_ https://github.com/bloodwass/mixout/blob/48165168e0066a39c90252c1017e4428313f5a8c/module.py#L18, a mixout version of _torch.nn.Linear_. So if your model does not have _torch.nn.Linear_, you should make your own mixout module using https://github.com/bloodwass/mixout/blob/48165168e0066a39c90252c1017e4428313f5a8c/mixout.py#L11

You can see an example of applying mixout in https://github.com/bloodwass/mixout/blob/master/example.py 

For more details,
you first call all modules (instance and name) of your pretrained model
https://github.com/bloodwass/mixout/blob/48165168e0066a39c90252c1017e4428313f5a8c/example.py#L39
I strongly recommend you use mixout without dropout, so you have to turn off your dropout layers if your model has them as follows:
https://github.com/bloodwass/mixout/blob/48165168e0066a39c90252c1017e4428313f5a8c/example.py#L40-L41

From now on, you apply mixout to your _torch.nn.Linear_ by replacing this linear module with _MixLinear_:
https://github.com/bloodwass/mixout/blob/48165168e0066a39c90252c1017e4428313f5a8c/example.py#L42-L48
In line 42, check whether your current module is _torch.nn.Linear_ or not.
If so, save your pretrained parameters (state_dict) of this (just like line 43).
After that, you construct a _MixLinear_ instance matching your original _nn.Module_. (See line 44-46.) Note that you must choose your mix probability and target parameters for mixing when designing the _MixLinear_ instance. (In my example code, they are '0.9' and 'target_state_dict['weight']', respectively).
At this point, your _MixLinear_ module has randomly initialized. It means that you need to load pre-trained parameters for this new module. Finally, you are ready for the substitution and do it by setattr (line 48). 

In summary,
1. turn off your dropout layer.
2. Find a linear module of your model.
3. With a mix probability and target parameters (usually pretrained parameters), construct a _MixLinear_ which has exactly same size of your original linear module (in_features, out_features, bias=True, etc.)
4. Initialize your new mixout layer with pretrained parameters
5. Substitute the mixout layer for the original linear layer by setattr
Thank you for your attention to mixout!
